courses/models.py

```diff
@@ -237,17 +215,52 @@ def is_upgradable(self):
         return (self.upgrade_deadline is None or
                 (self.upgrade_deadline > datetime.now(pytz.utc)))
 
-    @staticmethod
-    def unexpired_courserun_queryset():
+    @classmethod
+    def get_first_unexpired_run_qset(cls):
         """
-        Returns a Q object for CourseRuns that have not expired (ie: the enrollment
-        period has not passed, or the enrollment period is unspecified and the end
-        date occurs in the future)
+        It generates a queryset for fetching the first available run for a course.
         """
         now = datetime.now(pytz.utc)
-        return (
-            (models.Q(enrollment_end=None) & (
-                models.Q(end_date__gte=now) | models.Q(end_date=None)
+        unexpired_courserun_q = (
+            (models.Q(enrollment_end__isnull=True) & (
+                models.Q(end_date__gte=now) | models.Q(end_date__isnull=True)
             )) |
             models.Q(enrollment_end__gte=now)
         )
+        past_courserun_q = (
+            models.Q(end_date__isnull=False) &
+            models.Q(end_date__lt=now)
+        )
+        upgradable_q = (
+            models.Q(upgrade_deadline__isnull=True) |
+            models.Q(upgrade_deadline__gte=now)
+        )
+
+        return (
+            cls.objects
+            .filter(unexpired_courserun_q)
+            .exclude(past_courserun_q)
+            .filter(upgradable_q)
+            .order_by('start_date')
+            .select_related('course')
+        )
+
+    @classmethod
+    def get_first_unexpired_run(cls, course):
+        """
+        Gets the soonest unexpired run associated with the provided course.
+        Note that it could be current as long as the enrollment window
+        has not closed.
+
+        Args:
+            course (Course): A Course to look for in the query
+
+        Returns: CourseRun or None: An unexpired course run
+
+        """
+        future_run_queryset = cls.get_first_unexpired_run_qset()
+        # the reason for looping and not calling .first() is that
+        # in this way we avoid to do a new query for each course passed
+        for run in future_run_queryset:
+            if run.course.id == course.id:
```

Comment:
Wouldn't it be more efficient to set another filter on the queryset and let the database find the right CourseRun, rather than letting the database return _all_ of them and filter in the application logic?
```python
def get_first_unexpired_run(cls, course):
    query = cls.get_first_unexpired_run_qset()
    return query.filter(course_id=course.id).first()
```